### PR TITLE
Oblique Mercator class

### LINKED ISSUE #1618: Add Oblique Mercator
Will there be support for the oblique Mercator projection in Cartopy? I'd like to use it to plot some line-style geographic features that are sometimes best aligned with a rotation. Thanks!

_Originally posted by @shipengcheng1230 in https://github.com/SciTools/cartopy/issues/501#issuecomment-657857834_

### PR DESCRIPTION
**Pull request by trexfeathers**
## Rationale

Iris users are requesting (SciTools/iris#4830) support for the Oblique Mercator projection, in line with the CF Conventions. This will allow them to quickly and easily plot data from NetCDF files that use this projection.

* [Oblique Mercator in CF Conventions](https://cfconventions.org/Data/cf-conventions/cf-conventions-1.7/cf-conventions.html#_oblique_mercator)
* [Oblique Mercator in PROJ](https://proj.org/operations/projections/omerc.html)

(Code largely copied from `TransverseMercator`, with appropriate modifications).

## Implications

New feature, no other expected implications. Closes #1618.

To-do:

- [x] Tests
- [x] An entry in [`docs/source/reference/projections.rst`](https://github.com/SciTools/cartopy/blob/3d2f67321438b13207b5e5057715e11cbcb70d16/docs/source/reference/projections.rst)
- [x] Resolve the coastline 'folding' bug
  (https://github.com/SciTools/cartopy/pull/2096#issuecomment-1284192774)
- [x] Confirm whether the `x_limits` and `y_limits` settings are appropriate
  (https://github.com/SciTools/cartopy/pull/2096#discussion_r999621607)

## Usage example

```python
from cartopy import crs as ccrs
from matplotlib import pyplot as plt

projection = ccrs.ObliqueMercator()
ax = plt.axes(projection=projection)
ax.coastlines()
ax.gridlines()
plt.show()

```

oblique_mercator

### CODE REVIEW

File: lib/cartopy/crs.py [hunk 1]
```diff
@@ -3089,6 +3089,71 @@ def y_limits(self):
...
+                                 (x1, y1), (x1, y0),
+                                 (x0, y0)])
+
+    @property
+    def x_limits(self):
+        return (-2e7, 2e7)
+
+    @property
+    def y_limits(self):
+        return (-1e7, 1e7)
```
Comment (trexfeathers) [inline lib/cartopy/crs.py:3154]: These were copied across from `TransverseMercator`. I have no way of knowing whether these values are appropriate for `ObliqueMercator`.

**Comment by dopplershift** [comment]
I think setting more appropriate values for `x_limits` and `y_limits` might help automatically block out the folding? Essentially those values are the valid ranges for those coordinates in projected space.

File: lib/cartopy/tests/crs/test_oblique_mercator.py [hunk 2]
```diff
@@ -0,0 +1,190 @@
...
+# licensing details.
+"""
+Tests for the Oblique Mercator projection.
+
+"""
+
+from copy import deepcopy
+from typing import Dict, List, NamedTuple, Tuple
+
+from matplotlib import pyplot as plt
```
[the remark this answers is not in the document]
Comment (trexfeathers) [reply lib/cartopy/tests/crs/test_oblique_mercator.py]: 528e2f1eade7237245db57094afd0bfe7826a068

File: lib/cartopy/tests/crs/test_oblique_mercator.py [hunk 3]
```diff
@@ -0,0 +1,190 @@
...
+            lonc="0.0",
+            lat_0="0.0",
+            k="1.0",
+            x_0="0.0",
+            y_0="0.0",
+            alpha="0.0",
+            units="m",
+        )
+        proj_kwargs_expected.update(inputs.proj_kwargs)
+        self.proj_params = {f"{k}={v}" for k, v in proj_kwargs_expected.items()}
```
[the remark this answers is not in the document]
Comment (trexfeathers) [reply lib/cartopy/tests/crs/test_oblique_mercator.py]: 528e2f1eade7237245db57094afd0bfe7826a068

File: lib/cartopy/tests/crs/test_oblique_mercator.py [hunk 4]
```diff
@@ -0,0 +1,190 @@
...
+def test_equality(oblique_variants):
+    """Check == and != operators of ccrs.ObliqueMercator."""
+    default, alt_1, alt_2 = oblique_variants
+    assert alt_1 == alt_2
+    assert alt_1 != default
+    assert hash(alt_1) != hash(default)
+    assert hash(alt_1) == hash(alt_2)
+
+
+@pytest.mark.parametrize("reverse_coord", [False, True], ids=["xy_order", "yx_order"])
```
[the remark this answers is not in the document]
Comment (trexfeathers) [reply lib/cartopy/tests/crs/test_oblique_mercator.py]: 528e2f1eade7237245db57094afd0bfe7826a068

File: lib/cartopy/tests/mpl/test_mpl_integration.py [hunk 5]
```diff
@@ -188,6 +188,10 @@ def test_simple_global():
     ccrs.SouthPolarStereo,
     pytest.param((ccrs.TransverseMercator, dict(approx=True)),
                  id='TransverseMercator'),
+    pytest.param((ccrs.ObliqueMercator, dict(azimuth=0.)),
+                 id='ObliqueMercator_default'),
+    pytest.param((ccrs.ObliqueMercator, dict(azimuth=90., central_latitude=-22)),
```
[the remark this answers is not in the document]
Comment (trexfeathers) [reply lib/cartopy/tests/mpl/test_mpl_integration.py]: 528e2f1eade7237245db57094afd0bfe7826a068

**Comment by rcomer** [comment]
@trexfeathers I think it wasn’t you https://github.com/pydata/pydata-sphinx-theme/issues/1226

**Comment by trexfeathers** [comment]
> @trexfeathers I think it wasn’t you [pydata/pydata-sphinx-theme#1226](https://github.com/pydata/pydata-sphinx-theme/issues/1226)

✔

File: lib/cartopy/tests/crs/test_oblique_mercator.py [hunk 6]
```diff
@@ -0,0 +1,193 @@
...
+                src_crs=plate_carree,
+            )
+            np.testing.assert_array_almost_equal(res, expected, decimal=4)
+
+
+@pytest.fixture
+def oblique_variants(
+    oblique_mercator,
+    rotated_mercator,
+) -> Tuple[ccrs.ObliqueMercator, ccrs.ObliqueMercator, ccrs.ObliqueMercator]:
```
Comment (bjlittle) [inline lib/cartopy/tests/crs/test_oblique_mercator.py:168]: @trexfeathers Minor point, but didn't know if you wanted to create a [typing alias](https://docs.python.org/3/library/typing.html#type-aliases) for this?

Admittedly, this is a bit of a marmite comment. So happy for you to say "nope" :+1:

File: lib/cartopy/tests/crs/test_oblique_mercator.py [hunk 7]
```diff
@@ -0,0 +1,193 @@
...
+class TestCrsArgs:
+    point_a_plate_carree = (-3.474083, 50.727301)
+    point_b_plate_carree = (0.5, 50.5)
+
+    class ParamTuple(NamedTuple):
+        id: str
+        crs_kwargs: dict
+        proj_kwargs: Dict[str, str]
+        expected_a: Tuple[float, float]
+        expected_b: Tuple[float, float]
```
Comment (bjlittle) [inline lib/cartopy/tests/crs/test_oblique_mercator.py:45]: @trexfeathers Nice pattern :100:

File: lib/cartopy/tests/crs/test_oblique_mercator.py [hunk 8]
```diff
@@ -0,0 +1,193 @@
...
+def rotated_mercator() -> ccrs.ObliqueMercator:
+    return ccrs.ObliqueMercator(azimuth=90.0)
+
+
+@pytest.fixture
+def plate_carree() -> ccrs.PlateCarree:
+    return ccrs.PlateCarree()
+
+
+class TestCrsArgs:
```
Comment (bjlittle) [inline lib/cartopy/tests/crs/test_oblique_mercator.py:36]: @trexfeathers Minor, now, is it `TestCRSArgs` or `TestCrsArgs` :wink:

File: lib/cartopy/tests/crs/test_oblique_mercator.py [hunk 9]
```diff
@@ -0,0 +1,193 @@
...
+        proj_kwargs_expected = dict(
+            ellps="WGS84",
+            lonc="0.0",
+            lat_0="0.0",
+            k="1.0",
+            x_0="0.0",
+            y_0="0.0",
+            alpha="0.0",
+            units="m",
+        )
```
Comment (bjlittle) [inline lib/cartopy/tests/crs/test_oblique_mercator.py:138]: @trexfeathers Minor, you could move this to be a class attribute (akin to `TestCrsArgs.point_a_plate_carree` and `TestCrsArgs.point_b_plate_carree`) and save the exact same `dict` from being recreated each call to this fixture.

**Review by bjlittle — CHANGES_REQUESTED** [review]
@trexfeathers Thanks for this awesome contribution, much appreciated.

Just some super minor comments to service, and then we can bank this new CRS loveliness :smile:

File: lib/cartopy/tests/crs/test_oblique_mercator.py (hunk 8 above)
Comment (trexfeathers) [reply lib/cartopy/tests/crs/test_oblique_mercator.py:36]: This was a conscious choice - I struggle to read camel case when there's an acronym in there.

Based on [this Stack Overflow thread](https://stackoverflow.com/questions/15526107/acronyms-in-camelcase) it's clearly quite controversial, but there is majority support for `TestCrsArgs`.

File: lib/cartopy/tests/crs/test_oblique_mercator.py (hunk 6 above)
Comment (trexfeathers) [reply lib/cartopy/tests/crs/test_oblique_mercator.py:168]: Well it's certainly good to know that's an option. I know what I've written is ugly, but I think an alias in this case would be another type of ugly! So, "nope" 😄

File: lib/cartopy/tests/crs/test_oblique_mercator.py (hunk 9 above)
Comment (trexfeathers) [reply lib/cartopy/tests/crs/test_oblique_mercator.py]: 7ec343b9dade3571a8ce90d482cbf3d533e2aec9

File: lib/cartopy/crs.py [hunk 10]
```diff
@@ -3089,6 +3089,86 @@ def y_limits(self):
...
+        false_easting: optional
+            X offset from the planar origin in metres. Defaults to 0.
+        false_northing: optional
+            Y offset from the planar origin in metres. Defaults to 0.
+        scale_factor: optional
+            Scale factor at the central meridian. Defaults to 1.
+        azimuth: optional
+            Azimuth of centerline clockwise from north at the center point of
+            the centre line. Defaults to 0.
+
```
[the remark this answers is not in the document]
Comment (trexfeathers) [reply lib/cartopy/crs.py]: 7ec343b9dade3571a8ce90d482cbf3d533e2aec9

**Comment by bjlittle** [comment]
@trexfeathers Thanks again... now it's time to oblique my mercator!

Lovely contribution, cheers :beers:

**Comment by QuLogic** [comment]
[`ObliqueMercator` is not in the projections list](https://output.circle-artifacts.com/output/job/c3bc17ae-f24d-473a-9de3-655c478c2ef8/artifacts/0/docs/build/html/reference/projections.html#cartopy-projections) because you forgot to run `make projection`.

**Comment by bjlittle** [comment]
@QuLogic Thanks, see https://github.com/SciTools/cartopy/pull/2146 :+1:

**Comment by trexfeathers** [comment]
Forgot to say: thanks to @ESadek-MO and @scottrobinson02 for their help on this project ❤

**Pull request merged**